codegen/src/main/java/software/amazon/awssdk/codegen/poet/auth/scheme/ModelAuthSchemeKnowledgeIndex.java

```diff
@@ -105,17 +110,176 @@ private Map<List<String>, List<AuthTrait>> operationsToAuthOptions() {
                                       Map.Entry::getKey, (a, b) -> b,
                                       LinkedHashMap::new));
 
-        List<AuthTrait> serviceDefaults = serviceDefaultAuthOption();
+        List<AuthTrait> serviceDefaults = serviceDefaultAuthTrait();
 
-        // Handle operations with defaults
+        // Get the list of operations that share the same auth schemes as the system defaults and remove it from the result. We
+        // will take care of all of these in the fallback `default` case.
         List<String> operationsWithDefaults = authSchemesToOperations.remove(serviceDefaults);
         if (operationsWithDefaults != null) {
             operationsToAuthTrait.remove(operationsWithDefaults);
         }
-        operationsToAuthTrait.put(Collections.emptyList(), serviceDefaults);
+
         return operationsToAuthTrait;
     }
 
+    /**
+     * Determines if an operation should be included based on whether it has auth traits.
+     *
+     * <p>Returns false if:
+     * <ul>
+     *   <li>The operation has no modeled auth traits (empty auth list)</li>
+     * </ul>
+     *
+     * <p>Throws UnsupportedOperationException if:
+     * <ul>
+     *   <li>The service has hardcoded auth scheme overrides defined in {@link SigV4SignerDefaults}. The
+     *   exception is WriteGetObjectResponse, which has "v4-unsigned-body" auth type. It is already handled by the
+     *   S3 default auth schemes, so there is no need to add operation override</li>
+     * </ul>
+     */
+    private boolean hasAuthTrait(Map.Entry<String, OperationModel> op) {
+        if (op.getValue().getAuth().isEmpty()) {
+            return false;
+        }
+
+        if (!hasCustomServiceAuthSchemeOverride()) {
+            return true;
+        }
+
+        if (op.getKey().equals("WriteGetObjectResponse")) {
+            return false;
+        }
+
+        throw new UnsupportedOperationException(
+            String.format("Operation %s has auth trait and requires special handling: ", op.getKey()));
+    }
+
+    /**
+     * Returns a map from a list of operations to the list of auth-types modeled for those operations. The {@link AuthTrait}
```

Comment:
I tried my best to explain what this method does here because it's quite hard to understand, lmk if it's still not clear.

Reply:
FWIW, I think the explanation is good.